File: sdk/ts/test/modules.test.ts
// Mock-dispatcher tests for ModulesClient (epic memql#4183): the three
// module-registry envelopes round-trip, wire optionals normalize into the
// SDK-owned value types, payload-level errorCode throws, and the secret
// contract survives normalization -- a secret env var's value stays ""
// even when a (hypothetically misbehaving) server sent one, because the
// UI must never have a value to render for a secret.

import test from "node:test";
import assert from "node:assert/strict";

import { ModulesClient } from "../src/client/modules.js";
import type { Dispatcher } from "../src/client/dispatcher.js";
import type { ClientMessage, ServerMessage } from "../src/client/wire.js";

class MockDispatcher {
  readonly sent: Array<{ msg: ClientMessage; messageId: string }> = [];
  private pendingReplies = new Map<string, (msg: ServerMessage) => void>();
  private nextId = 0;

  send(msg: ClientMessage): string {
    const id = msg.messageId ?? `mock-${this.nextId++}`;
    this.sent.push({ msg: { ...msg, messageId: id }, messageId: id });
    return id;
  }

  async sendAndWait(msg: ClientMessage, signal?: AbortSignal): Promise<ServerMessage> {
    const id = msg.messageId ?? `mock-${this.nextId++}`;
    this.sent.push({ msg: { ...msg, messageId: id }, messageId: id });
    return new Promise<ServerMessage>((resolve, reject) => {
      if (signal?.aborted) {
        reject(new Error("aborted"));
        return;
      }
      this.pendingReplies.set(id, resolve);
    });
  }

  addEventListener(_handler: (msg: ServerMessage) => void): () => void {
    return () => {};
  }

  registerStream(_requestId: string, _handler: (msg: ServerMessage) => void): () => void {
    return () => {};
  }

  reply(payload: Record<string, unknown>): void {
    const last = this.sent.at(-1);
    if (!last) throw new Error("MockDispatcher.reply: nothing sent yet");
    const resolver = this.pendingReplies.get(last.messageId);
    if (!resolver) throw new Error(`MockDispatcher.reply: no pending entry for ${last.messageId}`);
    this.pendingReplies.delete(last.messageId);
    resolver({ correlateTo: last.messageId, ...payload } as ServerMessage);
  }

  asDispatcher(): Dispatcher {
    return this as unknown as Dispatcher;
  }
}

test("listModules -- sends modulesList and normalizes rows", async () => {
  const mock = new MockDispatcher();
  const mc = new ModulesClient(mock.asDispatcher());

  const promise = mc.listModules();
  const sent = mock.sent.at(-1)!.msg as unknown as { modulesList?: { requestId?: string } };
  assert.ok(sent.modulesList, "sends a modulesList payload");
  assert.ok(sent.modulesList.requestId, "carries a request id");

  mock.reply({
    modulesListResult: {
      modules: [
        {
          kind: "pack",
          name: "harness",
          state: "enabled",
          scope: "cluster",
          fqnPrefixes: ["integration.harnessRecall.", "integration.harnessTrace."],
        },
        { kind: "component", name: "identity" },
      ],
      reportingNodeId: "bff-abc",
      reportingNodeType: "bff",
    },
  });

  const inv = await promise;
  assert.equal(inv.modules.length, 2);
  assert.equal(inv.modules[0]!.name, "harness");
  assert.equal(inv.modules[0]!.scope, "cluster");
  assert.deepEqual(inv.modules[1]!.fqnPrefixes, []);
  assert.equal(inv.modules[1]!.stateDetail, "");
  assert.equal(inv.reportingNodeId, "bff-abc");
  assert.equal(inv.reportingNodeType, "bff");
});

test("listModules -- payload-level error throws with the engine's message", async () => {
  const mock = new MockDispatcher();
  const mc = new ModulesClient(mock.asDispatcher());

  const promise = mc.listModules();
  mock.reply({
    modulesListResult: { errorCode: 7, errorMessage: "modules: requires the owner or admin role" },
  });

  await assert.rejects(promise, /requires the owner or admin role/);
});

test("getModuleDetail -- sends (kind, name) and normalizes env vars", async () => {
  const mock = new MockDispatcher();
  const mc = new ModulesClient(mock.asDispatcher());

  const promise = mc.getModuleDetail("integration", "email");
  const sent = mock.sent.at(-1)!.msg as unknown as {
    moduleDetail?: { kind?: string; name?: string };
  };
  assert.equal(sent.moduleDetail?.kind, "integration");
  assert.equal(sent.moduleDetail?.name, "email");

  mock.reply({
    moduleDetailResult: {
      module: { kind: "integration", name: "email", state: "active" },
      envVars: [
        { name: "SMTP_HOST", secret: false, set: true, value: "mail.example.test" },
        { name: "SMTP_PASSWORD", secret: true, set: true },
      ],
      reportingNodeId: "bff-abc",
      reportingNodeType: "bff",
    },
  });

  const detail = await promise;
  assert.equal(detail.module.name, "email");
  assert.equal(detail.envVars.length, 2);
  assert.equal(detail.envVars[0]!.value, "mail.example.test");
  const secret = detail.envVars[1]!;
  assert.equal(secret.secret, true);
  assert.equal(secret.set, true);
  assert.equal(secret.value, "", "a secret entry never carries a value");
});

test("setPackEnabled -- round-trips the flip and keeps restartRequired honest", async () => {
  const mock = new MockDispatcher();
  const mc = new ModulesClient(mock.asDispatcher());

  const promise = mc.setPackEnabled("harness", false, "maintenance window");
  const sent = mock.sent.at(-1)!.msg as unknown as {
    setPackEnabled?: { packDomain?: string; enabled?: boolean; reason?: string };
  };
  assert.equal(sent.setPackEnabled?.packDomain, "harness");
  assert.equal(sent.setPackEnabled?.enabled, false);
  assert.equal(sent.setPackEnabled?.reason, "maintenance window");

  mock.reply({
    setPackEnabledResult: {
      packDomain: "harness",
      priorEnabled: true,
      enabled: false,
      restartRequired: true,
    },
  });

  const outcome = await promise;
  assert.equal(outcome.priorEnabled, true);
  assert.equal(outcome.enabled, false);
  assert.equal(outcome.restartRequired, true);
});

test("setPackEnabled -- owner refusal throws", async () => {
  const mock = new MockDispatcher();
  const mc = new ModulesClient(mock.asDispatcher());

  const promise = mc.setPackEnabled("harness", false, "");
  mock.reply({
    setPackEnabledResult: { errorCode: 7, errorMessage: "set pack enabled: owner only" },
  });

  await assert.rejects(promise, /owner only/);
});

File: sdk/go/modules/modules_test.go
package modules

import (
	"context"
	"errors"
	"testing"
	"time"

	"google.golang.org/grpc/metadata"

	memqlv1 "github.com/znasllc-io/memql/component/grpc/gen"

	"github.com/znasllc-io/memql/sdk/go/client"
)

// mockStream mirrors sdk/go/pack's scripted-server mock.
type mockStream struct {
	sendCh chan *memqlv1.MemqlClientMessage
	recvCh chan *memqlv1.MemqlServerMessage
}

func newMockStream() *mockStream {
	return &mockStream{
		sendCh: make(chan *memqlv1.MemqlClientMessage, 10),
		recvCh: make(chan *memqlv1.MemqlServerMessage, 10),
	}
}

func (m *mockStream) Send(msg *memqlv1.MemqlClientMessage) error { m.sendCh <- msg; return nil }
func (m *mockStream) Recv() (*memqlv1.MemqlServerMessage, error) {
	msg, ok := <-m.recvCh
	if !ok {
		return nil, context.Canceled
	}
	return msg, nil
}
func (m *mockStream) Header() (metadata.MD, error) { return nil, nil }
func (m *mockStream) Trailer() metadata.MD         { return nil }
func (m *mockStream) CloseSend() error             { close(m.recvCh); return nil }
func (m *mockStream) Context() context.Context     { return context.Background() }
func (m *mockStream) SendMsg(any) error            { return nil }
func (m *mockStream) RecvMsg(any) error            { return nil }

func reply(t *testing.T, stream *mockStream, fn func(*memqlv1.MemqlClientMessage) *memqlv1.MemqlServerMessage) {
	t.Helper()
	go func() {
		sent := <-stream.sendCh
		resp := fn(sent)
		resp.CorrelateTo = sent.GetMessageId()
		stream.recvCh <- resp
	}()
}

func newTestClient(t *testing.T) (*Client, *mockStream, func()) {
	t.Helper()
	stream := newMockStream()
	d := client.NewDispatcher(stream, nil)
	go d.Run()
	return NewClient(d), stream, d.Stop
}

func TestList(t *testing.T) {
	c, stream, stop := newTestClient(t)
	defer stop()

	reply(t, stream, func(req *memqlv1.MemqlClientMessage) *memqlv1.MemqlServerMessage {
		if req.GetModulesList() == nil {
			t.Errorf("expected ModulesListMsg, got %T", req.GetPayload())
		}
		return &memqlv1.MemqlServerMessage{
			Payload: &memqlv1.MemqlServerMessage_ModulesListResult{
				ModulesListResult: &memqlv1.ModulesListResult{
					ReportingNodeId:   "bff-abc",
					ReportingNodeType: "bff",
					Modules: []*memqlv1.ModuleInfo{
						{Kind: "pack", Name: "harness", State: "enabled", Scope: "cluster",
							FqnPrefixes: []string{"integration.harnessRecall."}},
						{Kind: "integration", Name: "email", State: "active", Scope: "node"},
					},
				},
			},
		}
	})

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	inv, err := c.List(ctx)
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if inv.ReportingNodeID != "bff-abc" || inv.ReportingNodeType != "bff" {
		t.Errorf("reporting node mismatch: %+v", inv)
	}
	if len(inv.Modules) != 2 || inv.Modules[0].Name != "harness" || inv.Modules[0].FqnPrefixes[0] != "integration.harnessRecall." {
		t.Errorf("modules mismatch: %+v", inv.Modules)
	}
}

func TestListRefusalSurfacesAsError(t *testing.T) {
	c, stream, stop := newTestClient(t)
	defer stop()

	reply(t, stream, func(*memqlv1.MemqlClientMessage) *memqlv1.MemqlServerMessage {
		return &memqlv1.MemqlServerMessage{
			Payload: &memqlv1.MemqlServerMessage_ModulesListResult{
				ModulesListResult: &memqlv1.ModulesListResult{
					ErrorCode:    7,
					ErrorMessage: "module inventory: reading modules requires the owner or admin role",
				},
			},
		}
	})

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	_, err := c.List(ctx)
	var refused *RefusedError
	if !errors.As(err, &refused) || refused.Code != 7 {
		t.Fatalf("expected RefusedError code 7, got %v", err)
	}
}

func TestGetCarriesEnvSurface(t *testing.T) {
	c, stream, stop := newTestClient(t)
	defer stop()

	reply(t, stream, func(req *memqlv1.MemqlClientMessage) *memqlv1.MemqlServerMessage {
		d := req.GetModuleDetail()
		if d.GetKind() != "component" || d.GetName() != "ai" {
			t.Errorf("detail request mismatch: %+v", d)
		}
		return &memqlv1.MemqlServerMessage{
			Payload: &memqlv1.MemqlServerMessage_ModuleDetailResult{
				ModuleDetailResult: &memqlv1.ModuleDetailResult{
					Module: &memqlv1.ModuleInfo{Kind: "component", Name: "ai", State: "built_in"},
					EnvVars: []*memqlv1.ModuleEnvVar{
						{Name: "MEMQL_OPENAI_API_KEY", Secret: true, Set: true},
						{Name: "MEMQL_OBSERVE_LEVEL", Secret: false, Set: true, Value: "count"},
					},
				},
			},
		}
	})

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	detail, err := c.Get(ctx, "component", "ai")
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if len(detail.EnvVars) != 2 {
		t.Fatalf("env vars: %+v", detail.EnvVars)
	}
	if secret := detail.EnvVars[0]; !secret.Secret || secret.Value != "" {
		t.Errorf("secret env var must carry set/unset and no value: %+v", secret)
	}
	if v := detail.EnvVars[1]; v.Value != "count" {
		t.Errorf("variable value lost: %+v", v)
	}
}

func TestSetPackEnabled(t *testing.T) {
	c, stream, stop := newTestClient(t)
	defer stop()

	reply(t, stream, func(req *memqlv1.MemqlClientMessage) *memqlv1.MemqlServerMessage {
		m := req.GetSetPackEnabled()
		if m.GetPackDomain() != "harness" || m.GetEnabled() {
			t.Errorf("flip request mismatch: %+v", m)
		}
		return &memqlv1.MemqlServerMessage{
			Payload: &memqlv1.MemqlServerMessage_SetPackEnabledResult{
				SetPackEnabledResult: &memqlv1.SetPackEnabledResult{
					PackDomain: "harness", PriorEnabled: true, Enabled: false, RestartRequired: true,
				},
			},
		}
	})

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	flip, err := c.SetPackEnabled(ctx, "harness", false, "maintenance")
	if err != nil {
		t.Fatalf("SetPackEnabled: %v", err)
	}
	if !flip.PriorEnabled || flip.Enabled || !flip.RestartRequired {
		t.Errorf("flip mismatch: %+v", flip)
	}
}
